### PR TITLE
Add a cache of MediaType objects

### PR DESCRIPTION
**Pull request by asoldano**

### CODE REVIEW

File: resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/delegates/MediaTypeHeaderDelegate.java [hunk 1]
```diff
@@ -7,17 +7,35 @@
...
+
+      if (map == null) return parse(type);
+      MediaType result = map.get(type);
+      if (result == null)
+      {
+         result = parse(type);
+         final int size = map.size();
+         if (size >= MAX_MT_CACHE_SIZE)
+         {
+            map.clear();
```
Comment (rsvoboda) [inline resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/delegates/MediaTypeHeaderDelegate.java]: Is full map clear really necessary ?

File: resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/delegates/MediaTypeHeaderDelegate.java [hunk 2]
```diff
@@ -7,17 +7,35 @@
...
+import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public class MediaTypeHeaderDelegate implements RuntimeDelegate.HeaderDelegate
 {
+   private static final int MAX_MT_CACHE_SIZE = Integer.getInteger("org.jboss.resteasy.plugins.delegates.max_mediatype_cache_size", 200);
+   private static final Map<String, MediaType> map = MAX_MT_CACHE_SIZE > 0 ? new ConcurrentHashMap<String, MediaType>() : null;
```
Comment (rsvoboda) [inline resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/delegates/MediaTypeHeaderDelegate.java]: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentHashMap.html#ConcurrentHashMap(int)

initial size could be set to the limit

File: resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/delegates/MediaTypeHeaderDelegate.java (hunk 1 above)
Comment (rsvoboda) [inline resteasy-jaxrs/src/main/java/org/jboss/resteasy/plugins/delegates/MediaTypeHeaderDelegate.java]: could be something like https://stackoverflow.com/a/36255034/6266477 used ?

**Comment by asoldano** [comment]
@rsvoboda I know full eviction of the cache is unfair and looks simplistic; in facts it's really just the most straightforward safety belt for avoiding explosion of the cache. A LRU approach would be better, assuming it's efficiently implemented. Let's see the benefits of the current solution first and then evaluate improvements, ok?

**Comment by rsvoboda** [comment]
+1 for simple solution now + evaluate improvements

**Comment by asoldano** [comment]
Closing for now as it looks like the improvements are not as great as expected. Might re-evaluate in future.

**Pull request closed without merging**